### DISCUSSION TITLE
JSON Compilation Database

### DISCUSSION THREAD

**Discussion by ivan-pi** [fortran-lang/fpm#960] [General]
I was scrolling through the product sheet of [Codee](https://www.codee.com/product/), a software product for automated code inspection.

Their [Datasheet (PDF, 540 KB)](https://codee.com/wp-content/uploads/2022/06/Codee-Product-Datasheet.pdf) mentions they support integration with fpm:

![image](https://github.com/fortran-lang/fpm/assets/21085643/75fcd8ac-26fe-4343-b7bc-1f8cdaf1017a)

The documentation for the CMake and Ninja options is here:
- CMake: https://cmake.org/cmake/help/latest/variable/CMAKE_EXPORT_COMPILE_COMMANDS.html
- Ninja: https://ninja-build.org/manual.html#_extra_tools (search for `compdb`)

It looks like the compilation database originated from the Clang/LLVM community: https://clang.llvm.org/docs/JSONCompilationDatabase.html

The tool bear mentioned for FPM is this one: https://github.com/rizsotto/Bear. It intercepts your compile commands somehow, and generates the JSON database. 

I wonder if this means, there are commercial users of Codee (and fpm), that asked for it.

**Comment by ivan-pi** [comment]
If I understand correctly, the idea is that your build system exports the `compilation_commands.json` file in the main project directory. Static analysis tools, e.g. based on `clang` (and `flang`, or `lfortran`), can then use this information to produce various reports. More info under [LibTooling](https://clang.llvm.org/docs/Tooling.html#libtooling).

Perhaps some of you are interested in this?

(cc @certik, @gnikit, @everythingfunctional, @rouson)

**Comment by everythingfunctional** [comment]
I have had a few conversations with the Codee developers. This is a functionality that would be useful for them. I was unaware of the bear tool, or that they had devised a workaround using it.

**Comment by manuelarenaz** [comment]
We have added additional public documentation in the following links:

https://docs.codee.com/faqs#what-tools-are-available-for-generating-the-compile_commandsjson-file

https://docs.codee.com/faqs#what-information-is-store-in-the-compile_commandsjson-file

Just in case you find it useful for FPM development roadmap.
